internal/controller/supervisorconfig/upstreamwatchers/upstream_watchers.go

```diff
@@ -260,75 +283,68 @@ func (g *GradatedConditions) Conditions() []*v1alpha1.Condition {
-
+func validateAndSetLDAPServerConnectivityAndSearchBase(ctx context.Context, validatedSecretVersionsCache ValidatedSettingsCacheI, upstream UpstreamGenericLDAPIDP, config *upstreamldap.ProviderConfig, currentSecretVersion string) (*v1alpha1.Condition, *v1alpha1.Condition) {
 	validatedSettings, hasPreviousValidatedSettings := validatedSecretVersionsCache.Get(upstream.Name(), currentSecretVersion, upstream.Generation())
 	var ldapConnectionValidCondition, searchBaseFoundCondition *v1alpha1.Condition
-	if !hasPreviousValidatedSettings {
+
+	if hasPreviousValidatedSettings && validatedSettings.UserSearchBase != "" && validatedSettings.GroupSearchBase != "" {
+		// Found previously validated settings in the cache (which is also not missing search base fields), so use them.
+		config.ConnectionProtocol = validatedSettings.LDAPConnectionProtocol
+		config.UserSearch.Base = validatedSettings.UserSearchBase
+		config.GroupSearch.Base = validatedSettings.GroupSearchBase
+		ldapConnectionValidCondition = validatedSettings.ConnectionValidCondition.DeepCopy()
+		searchBaseFoundCondition = validatedSettings.SearchBaseFoundCondition.DeepCopy()
+	} else {
+		// Did not find previously validated settings in the cache, so probe the LDAP server.
 		testConnectionTimeout, cancelFunc := context.WithTimeout(ctx, probeLDAPTimeout)
 		defer cancelFunc()
-
 		ldapConnectionValidCondition = TestConnection(testConnectionTimeout, upstream.Spec().BindSecretName(), config, currentSecretVersion)
 
 		searchBaseTimeout, cancelFunc := context.WithTimeout(ctx, probeLDAPTimeout)
 		defer cancelFunc()
 		searchBaseFoundCondition = upstream.Spec().DetectAndSetSearchBase(searchBaseTimeout, config)
 
-		if ldapConnectionValidCondition.Status == v1alpha1.ConditionTrue {
-			// if it's nil, don't worry about the search base condition. But if it exists make sure the status is true.
-			if searchBaseFoundCondition == nil || (searchBaseFoundCondition.Status == v1alpha1.ConditionTrue) {
-				// Remember (in-memory for this pod) that the controller has successfully validated the LDAP provider
-				// using this version of the Secret. This is for performance reasons, to avoid attempting to connect to
-				// the LDAP server more than is needed. If the pod restarts, it will attempt this validation again.
-				validatedSettings.LDAPConnectionProtocol = config.ConnectionProtocol
-				validatedSettings.BindSecretResourceVersion = currentSecretVersion
-				validatedSettings.Generation = upstream.Generation()
-				validatedSettings.UserSearchBase = config.UserSearch.Base
-				validatedSettings.GroupSearchBase = config.GroupSearch.Base
-				validatedSecretVersionsCache.Set(upstream.Name(), currentSecretVersion, upstream.Generation(), validatedSettings)
-			}
+		// When there were no failures, write the newly validated settings to the cache.
+		// It's okay for the search base condition to be nil, but if it exists make sure it was not a failure.
```

Comment:
Maybe add something about why/when it might be nil e.g. `It's okay for the search base condition to be nill (since it's only used by Active Directory providers), but...`

Reply:
👍